internal/messaging/response_handler.go

```diff
@@ -1,147 +1,536 @@
+		addErrorToResponseHeader(response, fmt.Sprintf("error packing data: %v", err))
+		return true
+	}
+
+	msg := ethereum.CallMsg{
+		To:   &bookingTokenAddress,
+		Data: packedData,
+	}
+	result, err := h.ethClient.CallContract(context.Background(), msg, nil)
 	if err != nil {
-		addErrorToResponseHeader(response, fmt.Sprintf("error parsing buyer address: %v", err))
+		h.logger.Infof("Error calling contract: %v", err)
+		addErrorToResponseHeader(response, fmt.Sprintf("error calling contract: %v", err))
 		return true
 	}
-	price, err := strconv.Atoi(response.MintResponse.Price.Value)
+
+	// Unpack the result
+	var supplierName string
+	err = abi.UnpackIntoInterface(&supplierName, "getSupplierName", result)
 	if err != nil {
-		addErrorToResponseHeader(response, fmt.Sprintf("error parsing price value: %v", err))
+		h.logger.Infof("Error unpacking result: %v", err)
+		addErrorToResponseHeader(response, fmt.Sprintf("error unpacking result: %v", err))
 		return true
 	}
-	success, txID, err := h.tvmClient.SendTxAndWait(ctx, createNFTAction(owner, buyer, uint64(response.MintResponse.BuyableUntil.Seconds), uint64(price), response.MintResponse.MintId))
+
+	if supplierName != h.cfg.SupplierName {
+		h.logger.Debugf("Not registered with correct supplier name: %v != %v", supplierName, h.cfg.SupplierName)
+		h.logger.Debugf("Registerring with supplierName: %v", h.cfg.SupplierName)
+		err := register(h.ethClient, abi, h.pk.ToECDSA(), h.cfg.SupplierName, bookingTokenAddress)
+		if err != nil {
+			addErrorToResponseHeader(response, fmt.Sprintf("error registering supplier: %v", err))
+			return true
+		}
+	} else {
+		h.logger.Debugf("Supplier is already registered with supplierName: %v", supplierName)
+	}
+	// TODO @evlekht ensure that request.MintRequest.BuyerAddress is c-chain address format, not x/p/t chain
+	buyer := common.HexToAddress(request.MintRequest.BuyerAddress)
+
+	h.logger.Debugf("abi: %v", abi)
+
+	// TODO @evlekht unhardocoded, figure out what it is at all
+	// uri := "data:application/json;base64,eyJuYW1lIjoiQ2FtaW5vIE1lc3NlbmdlciBCb29raW5nVG9rZW4gVGVzdCJ9Cg=="
+
+	// Get a Token URI for the token.
+	jsonPlain, uri, err := createTokenURIforMintResponse(response.MintResponse)
+	if err != nil {
+		addErrorToResponseHeader(response, fmt.Sprintf("error creating token URI: %v", err))
+		return true
+	}
+
+	h.logger.Debugf("Token URI JSON: %s\n", jsonPlain)
+
+	if response.MintResponse.BuyableUntil == nil || response.MintResponse.BuyableUntil.Seconds == 0 {
+		response.MintResponse.BuyableUntil = timestamppb.New(time.Now().Add(time.Second * time.Duration(h.cfg.BuyableUntilDefault)))
+	}
+	// MINT TOKEN
+	txID, tokenID, err := mint(
+		h.ethClient,
+		bookingTokenAddress,
+		abi,
+		h.pk.ToECDSA(),
+		buyer,
+		uri,
+		big.NewInt(response.MintResponse.BuyableUntil.Seconds),
+	)
 	if err != nil {
 		errMessage := fmt.Sprintf("error minting NFT: %v", err)
 		if errors.Is(err, context.DeadlineExceeded) {
```

Comment:
is this condition still relevant? I don't see any context, that we pass to mint func
Also see buy func call

Reply:
`if errors.Is(err, context.DeadlineExceeded)`
Should we leave this for the refactoring phase?

Would it be better to extend the mint and buy to include the context?  

Do I get this right: If we don't do that, the execution time of the functions is not included within the total context execution time, which is checked to find out if the deadline has been exceeded?

Reply:
If we don't pass existing context, that this context deadline won't be used for request.
Though, code that creates and handls context should be improved, since its not doing job in separate goroutine, so ctx.Done() call make limited sense.
Anyway, you should at least pass existing context to mint/buy funcs, since u'r introducing this funcs. And we'll care about the rest in refactoring
And if-error.Is block is still relevant